File: p442m/find_duplicates.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var findDuplicates = function (nums) {
  const ans = [];
  for (let n of nums) {
    let key = Math.abs(n) - 1;
    if (nums[key] > 0) nums[key] *= -1;
    else ans.push(key + 1);
  }
  return ans;
};
// TEST
[
  [[4, 1, 2, 3], []],
  [[5, 1, 2, 3, 5], [5]],
  [
    [4, 3, 2, 7, 8, 2, 3, 1],
    [2, 3],
  ],
  [[1, 1, 2], [1]],
  [[1], []],
].forEach(([nums, expected]) => {
  const actual = findDuplicates(nums);
  console.log('Find duplicates in', nums, '->', actual);
  console.assert(actual.toString() === expected.toString());
});

File: p442m/find_duplicates.py
from typing import List


class Solution:
    def findDuplicates(self, nums: List[int]) -> List[int]:
        ans = []
        for n in nums:
            a = abs(n)
            if nums[a - 1] < 0:
                ans.append(a)
            else:
                nums[a - 1] *= -1
        return ans


# TESTS
for nums, expected in [
    ([4, 3, 2, 7, 8, 2, 3, 1], [2, 3]),
    ([1, 1, 2], [1]),
    ([1], []),
]:
    sol = Solution()
    actual = sol.findDuplicates(nums)
    print("Find duplicates in", nums, "->", actual)
    assert sorted(actual) == sorted(expected)
